subscriptions/subscriptions-impl/src/main/java/com/duckduckgo/subscriptions/impl/SubscriptionsManager.kt

```diff
@@ -365,6 +381,159 @@ class RealSubscriptionsManager @Inject constructor(
+                val account = authRepository.getAccount()
+                if (account == null) {
+                    logcat { "Subs: Cannot switch plan - no account found" }
+                    return@withContext false
+                }
+
+                // Validate the new plan exists
+                val availableOffers = getSubscriptionOffer()
+                val targetOffer = availableOffers.find { it.planId == planId && it.offerId == offerId }
+                if (targetOffer == null) {
+                    logcat { "Subs: Cannot switch plan - target plan not found: $planId with offer $offerId" }
+                    return@withContext false
+                }
+
+                // Launch Google Play billing flow for subscription update
+                logcat { "Subs: Launching subscription update flow for plan: $planId" }
+
+                // Launch the subscription update flow using PlayBillingManager
+                withContext(dispatcherProvider.main()) {
+                    playBillingManager.launchSubscriptionUpdate(
+                        activity = activity,
+                        newPlanId = planId,
+                        externalId = account.externalId,
+                        newOfferId = offerId,
+                        oldPurchaseToken = currentPurchaseToken,
+                        replacementMode = replacementMode,
+                    )
+                }
+
+                // Wait for the purchase to complete by observing the purchase state
+                var purchaseCompleted = false
+                var purchaseToken: String? = null
+
+                val purchaseStateJob = launch {
+                    playBillingManager.purchaseState.collect { state ->
+                        when (state) {
+                            is PurchaseState.Purchased -> {
+                                logcat { "Subs: Purchase completed successfully" }
+                                purchaseToken = state.purchaseToken
+                                purchaseCompleted = true
+                            }
+                            is PurchaseState.Canceled -> {
+                                logcat { "Subs: Purchase was canceled" }
+                                purchaseCompleted = true
+                            }
+                            is PurchaseState.InProgress -> {
+                                logcat { "Subs: Purchase in progress..." }
+                            }
+                        }
+                    }
+                }
+
+                // Wait for purchase completion with timeout
+                val timeoutMs = 60_000L // 60 seconds timeout
+                val startTime = System.currentTimeMillis()
+
+                while (!purchaseCompleted && (System.currentTimeMillis() - startTime) < timeoutMs) {
+                    delay(500) // Check every 500ms
+                }
+
+                purchaseStateJob.cancel()
+
+                val newPurchaseToken = purchaseToken
```

Comment:
We should generally avoid looping to check value when there are useful utilities for implementing timeouts. Here's an example:

```suggestion
                val newPurchaseToken = withTimeoutOrNull(timeout = 60.seconds) {
                    playBillingManager.purchaseState
                        .filterNot { it is PurchaseState.InProgress }
                        .map { (it as? PurchaseState.Purchased)?.purchaseToken }
                        .firstOrNull()
                }
```